codeintel: Update goals (#1924)
sourcegraph/about: handbook/engineering/code-intelligence/goals.md

**Milestones:**

1. Update pings and dashboards to accurately reflect this goal's success criteria.
1. 🔄 Support incrementally indexing source code to reduce the time to useful intelligence.
1. 🔄 Support automatically indexing source code to reduce the amount of effort required from users.
1. 🔄 lsif-go provides results for 80% of Go code intelligence operations for three companies.
1. 🔄 lsif-clang provides results for 80% of Go code intelligence operations for three companies.
1. 🔄 lsif-java provides results for 80% of Go code intelligence operations for three companies.
1. lsif-node provides results for 80% of Go code intelligence operations for three companies.
1. Integrate with build tools to reduce the friction of indexing monorepos.
    - Support Bazel
    - Support Buck
    - Support Pants

>     - Support OAO

Reviewer: Did we conclude that we aren't including this?

Author: Just OAO in particular?

Reviewer: whoops yea, just OAO

> # Code Intelligence goals and priorities
> 
> ## Goals

Reviewer 2: ```suggestion
# Code Intelligence goals
```

**Milestones:**

> 1. Validate Go support: Reach 80% precise code intelligence for Go within 3 organizations.

Reviewer 2: What does 80% mean here? We need this to be measurable in a dashboard. My proposal is that we need to have a concrete WAU number for each of these milestones (and the sum should add up to at least 10k).

Reviewer 2: In other words, answer the question: "how many WAU do we think each of these milestones will bring us"?

Author: > We need this to be measurable in a dashboard.

This _is_ measurable in a dashboard. We break out code intel actions between lsif and search based. If we only fall back 20% of the time we've met our success criteria.

Author: Is your suggestion to say that milestones contributes to WAU or each customer contributes to WAU? In the former case we need to know how in-demand a language is and in the latter we need to know what the number of potential users of Sg are at that company.

Reviewer 3: This would be measurable as "X% of Go code intel queries are precise at a certain customer" and are working to get this data on the dashboard.

Author: We currently have this data, but not broken down by language. We will have this broken down by the end of this milestone.

Reviewer 2: Maria's wording is clearer.
```suggestion
1. Validate Go support: 80% of Go code intel queries are precise within 3 customers.
```

But, even with this updated wording, it doesn't connect the dots for me.

What needs to be "validated"? Do we expect achieving this milestone with directly increase WAU at these customers (or is there some other reason we are doing this)? If WAU is the point of this milestone, then it seems like we should be able to estimate/forecast how much we expect each of these actions to impact WAU.

For example (I am just making up numbers):

- [Deliver to $U](https://github.com/sourcegraph/customer/issues/78) +1111 WAU
- [Deliver to $S](https://github.com/sourcegraph/customer/issues/77) +2222 WAU
- [Deliver to $C](https://github.com/sourcegraph/customer/issues/76) +3333 WAU

If we do this for all the milestones, then we can "show our work" on how we got to the top level goal (which is currently 10k, a made up number).

# Cloud goals and priorities

## Goals

Track our progress in our [Looker dashboard](https://sourcegraph.looker.com/dashboards/131).

> ### The majority of code intel results are precise (for languages we support fully)

Reviewer 4: I don't feel strongly about this number so feel free to change it, but I think it is important to be more clear than fuzzy. You can either say, "all code intel results", or set some high percentage we want to attain. Majority, or most of is not clear enough to communicate the vision/intention.

```suggestion
### 95% of all code intel results are precise for any fully supported language
```

Author: We actually discussed this and our only concern was that we want to explicit state that we're aware of a very, very long tail that we don't really care to tackle. The long tail is composed of edge cases such as very particular deployment setups used by only one company.

95% seems fine as long as we see some wiggle room in the success criteria here.

Reviewer 4: Ah, thanks that helps. In that case, I'd write it as explicitly as you stated. (Fill in the blank/tweak as needed):

```
### All code intel results are precise for any fully supported language

There is a long tail of edge cases for every language, and as described <where?>, we aspire to reach <some level> for every language. "All code intel results" is defined as all code intel results that meet this described level.
```

Author: We've consolidated the goals (and I don't think this applies anymore). Could you take another look?

> ### 95% of all code intel results are precise for any fully supported language

Reviewer 2: The problem with this goal is that we can achieve it without actually working on the right things. Even if we have 100% for all of these, we will have failed if the denominator is small. We need actual adoption and usage of LSIF. If we build the perfect precise code intelligence but barely anyone uses it, we have failed, and we need a goal that reflects that.

A WAU metric would make sense to me here (and the milestones don't need to change if we think that is exactly what we need to do to drive WAU).

Reviewer 2: WAU is nice because it encourages us to build things that are valuable to a broad audience (we will build LSIF Java before we build lsif lisp).

_Languages we fully support currently include Go, TypeScript, C/C++, and Java._ Languages will be added to this list as we make progress toward our second goal, listed below.

> It is imperative that we raise adoption *deeply* within a particular language and ecosystem. The goal of the code intelligence team is to provide fast and precise answers about a user's _universe of code_. This can only happen when that entire universe is indexed. We intend to provide an index for _all_ such universes of code by making progress on two fronts. First, we need to raise excitement for the opportunities adopting precise code intelligence makes possible. Once a user sees what is possible, they should have little reason not to put in the effort of adoption. Second, we need to reduce the friction felt during indexing (especially during adoption). Users tend to perceive only the work we ask them to do to set up precise code intelligence and do not perceive the full benefit of precise code intelligence. Reducing the friction of indexing is a critical step in growing support.

Reviewer 4: This is a great paragraph on what we need to do (and we should keep it), I'd also love to see an added paragraph about the value this brings to the user - the problem that we solve for them. Why would I, the user, care about making LSIF configuration easier? I don't care if I don't understand why precise code intel will change my life as a developer.

Reviewer 3: Added a short Problem section to address this.

```suggestion
**Problem:** By default, users get basic code intel results which are often incorrect for tokens with common names (such as Get). This is solved by enabling precise code intel, which gives users faster, 100% accurate results for a definition or reference for a symbol they hover over. 

It is imperative that we raise adoption *deeply* within a particular language and ecosystem. We want to provide fast and precise answers about a user's _universe of code_. This can only happen when that entire universe is indexed. We intend to provide an index for _all_ such universes of code by making progress on two fronts. First, we need to raise excitement for the opportunities adopting precise code intelligence makes possible. Once a user sees what is possible, they should have little reason not to put in the effort of adoption. Second, we need to reduce the friction felt during indexing (especially during adoption). Users tend to perceive only the work we ask them to do to set up precise code intelligence and do not perceive the full benefit of precise code intelligence. Reducing the friction of indexing is a critical step in growing support.
```

> **Milestones:**

Reviewer 4: Are these in order of operation? We need to accomplish the first before the second? (obviously some things will be done in parallel, but we've prioritized this list)?

Author: Not necessarily, these can basically _all_ be done in parallel (for this goal).

Reviewer 2: Even if they can be done in parallel, they should be ordered by importance (as if we couldn't do them in parallel).

Reviewer 3: Agreed they should be ordered by importance. Let's finish agreeing on what the goals/milestones should be first and order them right after that.

> 1. Update pings and dashboards to accurately reflect this goal's success criteria.

Reviewer 3: I think this one belonged to the second goal? 

```suggestion

```

> 1. 🔄 Support incrementally indexing source code to reduce the time to useful intelligence.

Reviewer 4: Consider rephrasing these around what they solve. For example, see my stubbed suggestion. Orient around what the experience of the user will be, and/or what I can expect as a user. Note: some items will be hard to reword, so do your best and we can iterate :) 

```suggestion
1. 🔄  The time to useful intelligence is reduced such that... within X minutes of uploading...
```

Reviewer 3: Suggested change of wording for other milestones to reflect this

1. 🔄 Support automatically indexing source code to reduce the amount of effort required from users.

> 1. 🔄 lsif-go provides results for 80% of Go code intelligence operations for three companies.

Reviewer 4: ```suggestion
1. 🔄  Reach 80% precise code intelligence for Go within 3 organizations.
```

    - https://github.com/sourcegraph/customer/issues/76
    - https://github.com/sourcegraph/customer/issues/77
    - https://github.com/sourcegraph/customer/issues/78

> 1. 🔄 lsif-clang provides results for 80% of Go code intelligence operations for three companies.

Reviewer 3: ```suggestion
1. 🔄  Validate C++ support: lsif-clang provides results for 80% of code intelligence operations for three companies.
```

    - https://github.com/sourcegraph/customer/issues/86
    - https://github.com/sourcegraph/customer/issues/87
    - https://github.com/sourcegraph/customer/issues/88

> 1. 🔄 lsif-java provides results for 80% of Go code intelligence operations for three companies.

Reviewer 3: ```suggestion
1. 🔄 Validate Java support: lsif-java provides results for 80% of code intelligence operations for three companies.
```

    - https://github.com/sourcegraph/customer/issues/79
    - https://github.com/sourcegraph/customer/issues/80
    - https://github.com/sourcegraph/customer/issues/81

> 1. lsif-node provides results for 80% of Go code intelligence operations for three companies.

Reviewer 3: ```suggestion
1. Validate Typescript/Javascript support: lsif-node provides results for 80% of code intelligence operations for three companies.
```

1. Integrate with build tools to reduce the friction of indexing monorepos.
    - Support Bazel
    - Support Buck
    - Support Pants

> ### Increase the number of fully supported languages

Reviewer 2: This isn't really a goal. It is an algorithm for determining the next goal of the team after the team completes goal #1.

Reviewer 3: Based on the change suggestion on goal #1, adding more language support items could also be considered part of the milestones for growing precise code intel adoption. @efritz what do you think?

> ### Increase the number of languages that we support fully

Reviewer 4: ```suggestion
### Increase the number of fully supported languages
```

> **Problem:** Our first goal increases adoption _deeply_ at companies whose core stack aligns with the set of languages we fully support. Unfortunately, this says nothing about enterprise customers who are _partially_ supported or _unsupportable_ because we lack support for a common language. There will always be a large pool of untapped users we can reach by adding support for additional languages.

Reviewer 4: This is closer to a problem statement, but can still be clarified a bit more. This focuses on the value to Sourcegraph, update to focus on value to the customer. Think about which languages would most impact the developer community (e.g. Python vs. OCaml) This helps us make prioritization decisions, which is what we want from our goals.

Reviewer 3: Upon reading and reflecting on the above feedback, I'm inclined to think that this should also be part of the "Growing adoption" problem. My trail is thought is: Having support for more languages -> support for more customers/better support for current customers -> growing LSIF adoption among customer users. 

@efritz @christinaforney does this make sense to you? If so I'd merge this part of the problem with our #1 goal problem section (and do the same with Milestones).

> **Milestones:**

Reviewer 4: As a sub section of these milestones, it would be helpful to list out the "next" languages we are considering. We can use this to solicit feedback, communicate to customers, and to explain why we're choosing to look at different languages over others.

Author: We *explicitly* don't want to do that until we have an automated way to update our backing data. We can make the data we have public but we'd have to anonymize it since it has notes about particular customers.

(The first milestone states) We want to add ping data that will help us collect this kind of data.
